// jira.js/dist/agile/client/agileClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgileClient = void 0;
const baseClient_1 = require("../../clients/baseClient");
const backlog_1 = require("../backlog");
const board_1 = require("../board");
const builds_1 = require("../builds");
const deployments_1 = require("../deployments");
const developmentInformation_1 = require("../developmentInformation");
const devopsComponents_1 = require("../devopsComponents");
const epic_1 = require("../epic");
const featureFlags_1 = require("../featureFlags");
const issue_1 = require("../issue");
const operations_1 = require("../operations");
const remoteLinks_1 = require("../remoteLinks");
const securityInformation_1 = require("../securityInformation");
const sprint_1 = require("../sprint");
class AgileClient extends baseClient_1.BaseClient {
    constructor() {
        super(...arguments);
        this.backlog = new backlog_1.Backlog(this);
        this.board = new board_1.Board(this);
        this.builds = new builds_1.Builds(this);
        this.deployments = new deployments_1.Deployments(this);
        this.developmentInformation = new developmentInformation_1.DevelopmentInformation(this);
        this.devopsComponents = new devopsComponents_1.DevopsComponents(this);
        this.epic = new epic_1.Epic(this);
        this.featureFlags = new featureFlags_1.FeatureFlags(this);
        this.issue = new issue_1.Issue(this);
        this.operations = new operations_1.Operations(this);
        this.remoteLinks = new remoteLinks_1.RemoteLinks(this);
        this.securityInformation = new securityInformation_1.SecurityInformation(this);
        this.sprint = new sprint_1.Sprint(this);
    }
}
exports.AgileClient = AgileClient;
