// jira.js/dist/createClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientType = void 0;
exports.createClient = createClient;
const agile_1 = require("./agile");
const clients_1 = require("./clients");
const serviceDesk_1 = require("./serviceDesk");
const version2_1 = require("./version2");
const version3_1 = require("./version3");
var ClientType;
(function (ClientType) {
    ClientType["Agile"] = "agile";
    ClientType["Version2"] = "version2";
    ClientType["Version3"] = "version3";
    ClientType["ServiceDesk"] = "serviceDesk";
})(ClientType || (exports.ClientType = ClientType = {}));
function createClient(clientType, config) {
    switch (clientType) {
        case ClientType.Agile:
            return new agile_1.AgileClient(config);
        case ClientType.Version2:
            return new version2_1.Version2Client(config);
        case ClientType.Version3:
            return new version3_1.Version3Client(config);
        case ClientType.ServiceDesk:
            return new serviceDesk_1.ServiceDeskClient(config);
        default:
            return new clients_1.BaseClient(config);
    }
}
