// jira.js/dist/serviceDesk/requestType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestType = void 0;
const tslib_1 = require("tslib");
class RequestType {
    constructor(client) {
        this.client = client;
    }
    getAllRequestTypes(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/servicedeskapi/requesttype',
                method: 'GET',
                headers: {
                    'X-ExperimentalApi': 'opt-in',
                },
                params: {
                    searchQuery: parameters === null || parameters === void 0 ? void 0 : parameters.searchQuery,
                    serviceDeskId: parameters === null || parameters === void 0 ? void 0 : parameters.serviceDeskId,
                    start: parameters === null || parameters === void 0 ? void 0 : parameters.start,
                    limit: parameters === null || parameters === void 0 ? void 0 : parameters.limit,
                    expand: parameters === null || parameters === void 0 ? void 0 : parameters.expand,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.RequestType = RequestType;
