// jira.js/dist/version2/announcementBanner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnnouncementBanner = void 0;
const tslib_1 = require("tslib");
class AnnouncementBanner {
    constructor(client) {
        this.client = client;
    }
    getBanner(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/announcementBanner',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    setBanner(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/announcementBanner',
                method: 'PUT',
                data: {
                    isDismissible: parameters.isDismissible,
                    isEnabled: parameters.isEnabled,
                    message: parameters.message,
                    visibility: parameters.visibility,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.AnnouncementBanner = AnnouncementBanner;
