// jira.js/dist/version2/appDataPolicies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppDataPolicies = void 0;
const tslib_1 = require("tslib");
class AppDataPolicies {
    constructor(client) {
        this.client = client;
    }
    getPolicy(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/data-policy',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getPolicies(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/data-policy/project',
                method: 'GET',
                params: {
                    ids: parameters === null || parameters === void 0 ? void 0 : parameters.ids,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.AppDataPolicies = AppDataPolicies;
