// jira.js/dist/version2/appMigration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppMigration = void 0;
const tslib_1 = require("tslib");
class AppMigration {
    constructor(client) {
        this.client = client;
    }
    updateIssueFields(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/atlassian-connect/1/migration/field',
                method: 'PUT',
                headers: {
                    'Atlassian-Account-Id': parameters.accountId,
                    'Atlassian-Transfer-Id': parameters.transferId,
                },
                data: {
                    updateValueList: parameters.updateValueList,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateEntityPropertiesValue(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/atlassian-connect/1/migration/properties/${parameters.entityType}`,
                method: 'PUT',
                headers: {
                    'Atlassian-Account-Id': parameters.accountId,
                    'Atlassian-Transfer-Id': parameters.transferId,
                    'Content-Type': 'application/json',
                },
                data: parameters.entities,
            };
            return this.client.sendRequest(config, callback);
        });
    }
    workflowRuleSearch(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/atlassian-connect/1/migration/workflow/rule/search',
                method: 'POST',
                headers: {
                    'Atlassian-Transfer-Id': parameters.transferId,
                },
                data: {
                    expand: parameters.expand,
                    ruleIds: parameters.ruleIds,
                    workflowEntityId: parameters.workflowEntityId,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.AppMigration = AppMigration;
