// jira.js/dist/version2/dynamicModules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamicModules = void 0;
const tslib_1 = require("tslib");
class DynamicModules {
    constructor(client) {
        this.client = client;
    }
    getModules(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/atlassian-connect/1/app/module/dynamic',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    registerModules(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/atlassian-connect/1/app/module/dynamic',
                method: 'POST',
                data: {
                    modules: parameters === null || parameters === void 0 ? void 0 : parameters.modules,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    removeModules(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/atlassian-connect/1/app/module/dynamic',
                method: 'DELETE',
                params: {
                    moduleKey: parameters === null || parameters === void 0 ? void 0 : parameters.moduleKey,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.DynamicModules = DynamicModules;
