// jira.js/dist/version2/filterSharing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FilterSharing = void 0;
const tslib_1 = require("tslib");
class FilterSharing {
    constructor(client) {
        this.client = client;
    }
    getDefaultShareScope(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/filter/defaultShareScope',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    setDefaultShareScope(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const scope = typeof parameters === 'string' ? parameters : parameters.scope;
            const config = {
                url: '/rest/api/2/filter/defaultShareScope',
                method: 'PUT',
                data: {
                    scope,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getSharePermissions(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/filter/${id}/permission`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    addSharePermission(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/filter/${parameters.id}/permission`,
                method: 'POST',
                data: {
                    type: parameters.type,
                    projectId: parameters.projectId,
                    groupname: parameters.groupname,
                    projectRoleId: parameters.projectRoleId,
                    accountId: parameters.accountId,
                    rights: parameters.rights,
                    groupId: parameters.groupId,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getSharePermission(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/filter/${parameters.id}/permission/${parameters.permissionId}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteSharePermission(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/filter/${parameters.id}/permission/${parameters.permissionId}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.FilterSharing = FilterSharing;
