// jira.js/dist/version2/issueSecurityLevel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueSecurityLevel = void 0;
const tslib_1 = require("tslib");
class IssueSecurityLevel {
    constructor(client) {
        this.client = client;
    }
    getIssueSecurityLevelMembers(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const issueSecuritySchemeId = typeof parameters === 'string' ? parameters : parameters.issueSecuritySchemeId;
            const config = {
                url: `/rest/api/2/issuesecurityschemes/${issueSecuritySchemeId}/members`,
                method: 'GET',
                params: {
                    startAt: typeof parameters !== 'string' && parameters.startAt,
                    maxResults: typeof parameters !== 'string' && parameters.maxResults,
                    issueSecurityLevelId: typeof parameters !== 'string' && parameters.issueSecurityLevelId,
                    expand: typeof parameters !== 'string' && parameters.expand,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getIssueSecurityLevel(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/securitylevel/${id}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.IssueSecurityLevel = IssueSecurityLevel;
