// jira.js/dist/version2/issues.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Issues = void 0;
const tslib_1 = require("tslib");
class Issues {
    constructor(client) {
        this.client = client;
    }
    getBulkChangelogs(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/changelog/bulkfetch',
                method: 'POST',
                data: {
                    fieldIds: parameters.fieldIds,
                    issueIdsOrKeys: parameters.issueIdsOrKeys,
                    maxResults: parameters.maxResults,
                    nextPageToken: parameters.nextPageToken,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getEvents(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/events',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    createIssue(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/issue',
                method: 'POST',
                params: {
                    updateHistory: parameters.updateHistory,
                },
                data: {
                    fields: parameters.fields,
                    historyMetadata: parameters.historyMetadata,
                    properties: parameters.properties,
                    transition: parameters.transition,
                    update: parameters.update,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    archiveIssuesAsync(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/issue/archive',
                method: 'POST',
                data: {
                    jql: parameters.jql,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    archiveIssues(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/issue/archive',
                method: 'PUT',
                data: {
                    issueIdsOrKeys: parameters.issueIdsOrKeys,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    createIssues(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/issue/bulk',
                method: 'POST',
                data: {
                    issueUpdates: parameters === null || parameters === void 0 ? void 0 : parameters.issueUpdates,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    bulkFetchIssues(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/issue/bulkfetch',
                method: 'POST',
                data: {
                    expand: parameters.expand,
                    fields: parameters.fields,
                    fieldsByKeys: parameters.fieldsByKeys,
                    issueIdsOrKeys: parameters.issueIdsOrKeys,
                    properties: parameters.properties,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getCreateIssueMeta(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/issue/createmeta',
                method: 'GET',
                params: {
                    projectIds: parameters === null || parameters === void 0 ? void 0 : parameters.projectIds,
                    projectKeys: parameters === null || parameters === void 0 ? void 0 : parameters.projectKeys,
                    issuetypeIds: parameters === null || parameters === void 0 ? void 0 : parameters.issuetypeIds,
                    issuetypeNames: parameters === null || parameters === void 0 ? void 0 : parameters.issuetypeNames,
                    expand: parameters === null || parameters === void 0 ? void 0 : parameters.expand,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getCreateIssueMetaIssueTypes(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/issue/createmeta/${parameters.projectIdOrKey}/issuetypes`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getCreateIssueMetaIssueTypeId(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/issue/createmeta/${parameters.projectIdOrKey}/issuetypes/${parameters.issueTypeId}`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getIssueLimitReport(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/issue/limit/report',
                method: 'GET',
                params: {
                    isReturningKeys: parameters === null || parameters === void 0 ? void 0 : parameters.isReturningKeys,
                },
                data: {
                    issuesApproachingLimitParams: parameters === null || parameters === void 0 ? void 0 : parameters.issuesApproachingLimitParams,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    unarchiveIssues(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/issue/unarchive',
                method: 'PUT',
                data: {
                    issueIdsOrKeys: parameters.issueIdsOrKeys,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getIssue(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const issueIdOrKey = typeof parameters === 'string' ? parameters : parameters.issueIdOrKey;
            const config = {
                url: `/rest/api/2/issue/${issueIdOrKey}`,
                method: 'GET',
                params: {
                    fields: typeof parameters !== 'string' ? parameters.fields : undefined,
                    fieldsByKeys: typeof parameters !== 'string' ? parameters.fieldsByKeys : undefined,
                    expand: typeof parameters !== 'string' ? parameters.expand : undefined,
                    properties: typeof parameters !== 'string' ? parameters.properties : undefined,
                    updateHistory: typeof parameters !== 'string' ? parameters.updateHistory : undefined,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    editIssue(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/issue/${parameters.issueIdOrKey}`,
                method: 'PUT',
                params: {
                    notifyUsers: parameters.notifyUsers,
                    overrideScreenSecurity: parameters.overrideScreenSecurity,
                    overrideEditableFlag: parameters.overrideEditableFlag,
                    returnIssue: parameters.returnIssue,
                    expand: parameters.expand,
                },
                data: {
                    fields: parameters.fields,
                    historyMetadata: parameters.historyMetadata,
                    properties: parameters.properties,
                    transition: parameters.transition,
                    update: parameters.update,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteIssue(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const issueIdOrKey = typeof parameters === 'string' ? parameters : parameters.issueIdOrKey;
            const config = {
                url: `/rest/api/2/issue/${issueIdOrKey}`,
                method: 'DELETE',
                params: {
                    deleteSubtasks: typeof parameters !== 'string' ? parameters.deleteSubtasks : undefined,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    assignIssue(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/issue/${parameters.issueIdOrKey}/assignee`,
                method: 'PUT',
                data: {
                    accountId: parameters.accountId,
                    accountType: parameters.accountType,
                    active: parameters.active,
                    applicationRoles: parameters.applicationRoles,
                    avatarUrls: parameters.avatarUrls,
                    displayName: parameters.displayName,
                    emailAddress: parameters.emailAddress,
                    expand: parameters.expand,
                    groups: parameters.groups,
                    key: parameters.key,
                    locale: parameters.locale,
                    name: parameters.name,
                    self: parameters.self,
                    timeZone: parameters.timeZone,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getChangeLogs(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const issueIdOrKey = typeof parameters === 'string' ? parameters : parameters.issueIdOrKey;
            const config = {
                url: `/rest/api/2/issue/${issueIdOrKey}/changelog`,
                method: 'GET',
                params: {
                    startAt: typeof parameters !== 'string' ? parameters.startAt : undefined,
                    maxResults: typeof parameters !== 'string' ? parameters.maxResults : undefined,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getChangeLogsByIds(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/issue/${parameters.issueIdOrKey}/changelog/list`,
                method: 'POST',
                data: {
                    changelogIds: parameters.changelogIds,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getEditIssueMeta(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const issueIdOrKey = typeof parameters === 'string' ? parameters : parameters.issueIdOrKey;
            const config = {
                url: `/rest/api/2/issue/${issueIdOrKey}/editmeta`,
                method: 'GET',
                params: {
                    overrideScreenSecurity: typeof parameters !== 'string' ? parameters.overrideScreenSecurity : undefined,
                    overrideEditableFlag: typeof parameters !== 'string' ? parameters.overrideEditableFlag : undefined,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    notify(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/issue/${parameters.issueIdOrKey}/notify`,
                method: 'POST',
                data: {
                    htmlBody: parameters.htmlBody,
                    restrict: parameters.restrict,
                    subject: parameters.subject,
                    textBody: parameters.textBody,
                    to: parameters.to,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getTransitions(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const issueIdOrKey = typeof parameters === 'string' ? parameters : parameters.issueIdOrKey;
            const config = {
                url: `/rest/api/2/issue/${issueIdOrKey}/transitions`,
                method: 'GET',
                params: {
                    expand: typeof parameters !== 'string' ? parameters.expand : undefined,
                    transitionId: typeof parameters !== 'string' ? parameters.transitionId : undefined,
                    skipRemoteOnlyCondition: typeof parameters !== 'string' ? parameters.skipRemoteOnlyCondition : undefined,
                    includeUnavailableTransitions: typeof parameters !== 'string' ? parameters.includeUnavailableTransitions : undefined,
                    sortByOpsBarAndStatus: typeof parameters !== 'string' ? parameters.sortByOpsBarAndStatus : undefined,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    doTransition(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/issue/${parameters.issueIdOrKey}/transitions`,
                method: 'POST',
                data: {
                    fields: parameters.fields,
                    historyMetadata: parameters.historyMetadata,
                    properties: parameters.properties,
                    transition: parameters.transition,
                    update: parameters.update,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    exportArchivedIssues(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/issues/archive/export',
                method: 'PUT',
                data: {
                    archivedBy: parameters.archivedBy,
                    archivedDateRange: parameters.archivedDateRange,
                    issueTypes: parameters.issueTypes,
                    projects: parameters.projects,
                    reporters: parameters.reporters,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.Issues = Issues;
