// jira.js/dist/version2/issueCustomFieldOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueCustomFieldOptions = void 0;
const tslib_1 = require("tslib");
class IssueCustomFieldOptions {
    constructor(client) {
        this.client = client;
    }
    getCustomFieldOption(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/customFieldOption/${id}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getOptionsForContext(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/field/${parameters.fieldId}/context/${parameters.contextId}/option`,
                method: 'GET',
                params: {
                    optionId: parameters.optionId,
                    onlyOptions: parameters.onlyOptions,
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    createCustomFieldOption(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/field/${parameters.fieldId}/context/${parameters.contextId}/option`,
                method: 'POST',
                data: {
                    options: parameters.options,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateCustomFieldOption(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/field/${parameters.fieldId}/context/${parameters.contextId}/option`,
                method: 'PUT',
                data: {
                    options: parameters.options,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    reorderCustomFieldOptions(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/field/${parameters.fieldId}/context/${parameters.contextId}/option/move`,
                method: 'PUT',
                data: {
                    after: parameters.after,
                    customFieldOptionIds: parameters.customFieldOptionIds,
                    position: parameters.position,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteCustomFieldOption(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/field/${parameters.fieldId}/context/${parameters.contextId}/option/${parameters.optionId}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    replaceCustomFieldOption(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/field/${parameters.fieldId}/context/${parameters.contextId}/option/${parameters.optionId}/issue`,
                method: 'DELETE',
                params: {
                    replaceWith: parameters.replaceWith,
                    jql: parameters.jql,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.IssueCustomFieldOptions = IssueCustomFieldOptions;
