// jira.js/dist/version2/issueLinks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueLinks = void 0;
const tslib_1 = require("tslib");
class IssueLinks {
    constructor(client) {
        this.client = client;
    }
    linkIssues(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/issueLink',
                method: 'POST',
                data: {
                    type: parameters.type,
                    inwardIssue: parameters.inwardIssue,
                    outwardIssue: parameters.outwardIssue,
                    comment: parameters.comment,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getIssueLink(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const linkId = typeof parameters === 'string' ? parameters : parameters.linkId;
            const config = {
                url: `/rest/api/2/issueLink/${linkId}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteIssueLink(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const linkId = typeof parameters === 'string' ? parameters : parameters.linkId;
            const config = {
                url: `/rest/api/2/issueLink/${linkId}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.IssueLinks = IssueLinks;
