// jira.js/dist/version2/issueSearch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueSearch = void 0;
const tslib_1 = require("tslib");
class IssueSearch {
    constructor(client) {
        this.client = client;
    }
    getIssuePickerResource(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/issue/picker',
                method: 'GET',
                params: {
                    query: parameters === null || parameters === void 0 ? void 0 : parameters.query,
                    currentJQL: parameters === null || parameters === void 0 ? void 0 : parameters.currentJQL,
                    currentIssueKey: parameters === null || parameters === void 0 ? void 0 : parameters.currentIssueKey,
                    currentProjectId: parameters === null || parameters === void 0 ? void 0 : parameters.currentProjectId,
                    showSubTasks: parameters === null || parameters === void 0 ? void 0 : parameters.showSubTasks,
                    showSubTaskParent: parameters === null || parameters === void 0 ? void 0 : parameters.showSubTaskParent,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    matchIssues(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/jql/match',
                method: 'POST',
                data: {
                    issueIds: parameters.issueIds,
                    jqls: parameters.jqls,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    searchForIssuesUsingJql(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/search',
                method: 'GET',
                params: {
                    jql: parameters.jql,
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                    validateQuery: parameters.validateQuery,
                    fields: parameters.fields,
                    expand: parameters.expand,
                    properties: parameters.properties,
                    fieldsByKeys: parameters.fieldsByKeys,
                    failFast: parameters.failFast,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    searchForIssuesUsingJqlPost(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/search',
                method: 'POST',
                data: {
                    expand: parameters === null || parameters === void 0 ? void 0 : parameters.expand,
                    fields: parameters === null || parameters === void 0 ? void 0 : parameters.fields,
                    fieldsByKeys: parameters === null || parameters === void 0 ? void 0 : parameters.fieldsByKeys,
                    jql: parameters === null || parameters === void 0 ? void 0 : parameters.jql,
                    maxResults: parameters === null || parameters === void 0 ? void 0 : parameters.maxResults,
                    properties: parameters === null || parameters === void 0 ? void 0 : parameters.properties,
                    startAt: parameters === null || parameters === void 0 ? void 0 : parameters.startAt,
                    validateQuery: parameters === null || parameters === void 0 ? void 0 : parameters.validateQuery,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    countIssues(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/search/approximate-count',
                method: 'POST',
                data: {
                    jql: parameters.jql,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    searchForIssuesIds(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/search/id',
                method: 'POST',
                data: {
                    jql: parameters.jql,
                    maxResults: parameters.maxResults,
                    nextPageToken: parameters.nextPageToken,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    searchForIssuesUsingJqlEnhancedSearch(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/search/jql',
                method: 'GET',
                params: {
                    jql: parameters.jql,
                    nextPageToken: parameters.nextPageToken,
                    maxResults: parameters.maxResults,
                    fields: parameters.fields,
                    expand: parameters.expand,
                    properties: parameters.properties,
                    fieldsByKeys: parameters.fieldsByKeys,
                    failFast: parameters.failFast,
                    reconcileIssues: parameters.reconcileIssues,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    searchForIssuesUsingJqlEnhancedSearchPost(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/search/jql',
                method: 'POST',
                data: {
                    jql: parameters.jql,
                    nextPageToken: parameters.nextPageToken,
                    maxResults: parameters.maxResults,
                    fields: parameters.fields,
                    expand: parameters.expand,
                    properties: parameters.properties,
                    fieldsByKeys: parameters.fieldsByKeys,
                    failFast: parameters.failFast,
                    reconcileIssues: parameters.reconcileIssues,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.IssueSearch = IssueSearch;
