// jira.js/dist/version2/issueTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueTypes = void 0;
const tslib_1 = require("tslib");
class IssueTypes {
    constructor(client) {
        this.client = client;
    }
    getIssueAllTypes(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/issuetype',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    createIssueType(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/issuetype',
                method: 'POST',
                data: {
                    description: parameters.description,
                    hierarchyLevel: parameters.hierarchyLevel,
                    name: parameters.name,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getIssueTypesForProject(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/issuetype/project',
                method: 'GET',
                params: {
                    projectId: parameters.projectId,
                    level: parameters.level,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getIssueType(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/issuetype/${id}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateIssueType(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/issuetype/${parameters.id}`,
                method: 'PUT',
                data: {
                    avatarId: parameters.avatarId,
                    description: parameters.description,
                    name: parameters.name,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteIssueType(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/issuetype/${id}`,
                method: 'DELETE',
                params: {
                    alternativeIssueTypeId: typeof parameters !== 'string' ? parameters.alternativeIssueTypeId : undefined,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getAlternativeIssueTypes(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/issuetype/${id}/alternatives`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    createIssueTypeAvatar(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            var _a;
            const config = {
                url: `/rest/api/2/issuetype/${parameters.id}/avatar2`,
                method: 'POST',
                headers: {
                    'X-Atlassian-Token': 'no-check',
                    'Content-Type': parameters.mimeType,
                },
                params: {
                    x: parameters.x,
                    y: parameters.y,
                    size: (_a = parameters.size) !== null && _a !== void 0 ? _a : 0,
                },
                data: parameters.avatar,
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.IssueTypes = IssueTypes;
