// jira.js/dist/version2/issueTypeProperties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueTypeProperties = void 0;
const tslib_1 = require("tslib");
class IssueTypeProperties {
    constructor(client) {
        this.client = client;
    }
    getIssueTypePropertyKeys(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const issueTypeId = typeof parameters === 'string' ? parameters : parameters.issueTypeId;
            const config = {
                url: `/rest/api/2/issuetype/${issueTypeId}/properties`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getIssueTypeProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/issuetype/${parameters.issueTypeId}/properties/${parameters.propertyKey}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    setIssueTypeProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/issuetype/${parameters.issueTypeId}/properties/${parameters.propertyKey}`,
                method: 'PUT',
                data: parameters.propertyValue,
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteIssueTypeProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/issuetype/${parameters.issueTypeId}/properties/${parameters.propertyKey}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.IssueTypeProperties = IssueTypeProperties;
