// jira.js/dist/version2/jiraSettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JiraSettings = void 0;
const tslib_1 = require("tslib");
class JiraSettings {
    constructor(client) {
        this.client = client;
    }
    getApplicationProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/application-properties',
                method: 'GET',
                params: {
                    key: parameters === null || parameters === void 0 ? void 0 : parameters.key,
                    permissionLevel: parameters === null || parameters === void 0 ? void 0 : parameters.permissionLevel,
                    keyFilter: parameters === null || parameters === void 0 ? void 0 : parameters.keyFilter,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getAdvancedSettings(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/application-properties/advanced-settings',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    setApplicationProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/application-properties/${parameters.id}`,
                method: 'PUT',
                data: parameters.body,
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getConfiguration(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/configuration',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.JiraSettings = JiraSettings;
