// jira.js/dist/version2/jQL.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JQL = void 0;
const tslib_1 = require("tslib");
class JQL {
    constructor(client) {
        this.client = client;
    }
    getAutoComplete(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/jql/autocompletedata',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getAutoCompletePost(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/jql/autocompletedata',
                method: 'POST',
                data: {
                    includeCollapsedFields: parameters === null || parameters === void 0 ? void 0 : parameters.includeCollapsedFields,
                    projectIds: parameters === null || parameters === void 0 ? void 0 : parameters.projectIds,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getFieldAutoCompleteForQueryString(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/jql/autocompletedata/suggestions',
                method: 'GET',
                params: {
                    fieldName: parameters === null || parameters === void 0 ? void 0 : parameters.fieldName,
                    fieldValue: parameters === null || parameters === void 0 ? void 0 : parameters.fieldValue,
                    predicateName: parameters === null || parameters === void 0 ? void 0 : parameters.predicateName,
                    predicateValue: parameters === null || parameters === void 0 ? void 0 : parameters.predicateValue,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    parseJqlQueries(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/jql/parse',
                method: 'POST',
                params: {
                    validation: parameters.validation,
                },
                data: {
                    queries: parameters.queries,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    migrateQueries(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/jql/pdcleaner',
                method: 'POST',
                data: {
                    queryStrings: parameters === null || parameters === void 0 ? void 0 : parameters.queryStrings,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    sanitiseJqlQueries(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/jql/sanitize',
                method: 'POST',
                data: {
                    queries: parameters === null || parameters === void 0 ? void 0 : parameters.queries,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.JQL = JQL;
