// jira.js/dist/version2/licenseMetrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LicenseMetrics = void 0;
const tslib_1 = require("tslib");
class LicenseMetrics {
    constructor(client) {
        this.client = client;
    }
    getLicense(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/instance/license',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getApproximateLicenseCount(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/license/approximateLicenseCount',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getApproximateApplicationLicenseCount(applicationKey, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/license/approximateLicenseCount/product/${applicationKey}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.LicenseMetrics = LicenseMetrics;
