// jira.js/dist/version2/projectKeyAndNameValidation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectKeyAndNameValidation = void 0;
const tslib_1 = require("tslib");
class ProjectKeyAndNameValidation {
    constructor(client) {
        this.client = client;
    }
    validateProjectKey(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const key = typeof parameters === 'string' ? parameters : parameters === null || parameters === void 0 ? void 0 : parameters.key;
            const config = {
                url: '/rest/api/2/projectvalidate/key',
                method: 'GET',
                params: {
                    key,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getValidProjectKey(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const key = typeof parameters === 'string' ? parameters : parameters === null || parameters === void 0 ? void 0 : parameters.key;
            const config = {
                url: '/rest/api/2/projectvalidate/validProjectKey',
                method: 'GET',
                params: {
                    key,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getValidProjectName(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const name = typeof parameters === 'string' ? parameters : parameters.name;
            const config = {
                url: '/rest/api/2/projectvalidate/validProjectName',
                method: 'GET',
                params: {
                    name,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.ProjectKeyAndNameValidation = ProjectKeyAndNameValidation;
