// jira.js/dist/version2/projectVersions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectVersions = void 0;
const tslib_1 = require("tslib");
class ProjectVersions {
    constructor(client) {
        this.client = client;
    }
    getProjectVersionsPaginated(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const projectIdOrKey = typeof parameters === 'string' ? parameters : parameters.projectIdOrKey;
            const config = {
                url: `/rest/api/2/project/${projectIdOrKey}/version`,
                method: 'GET',
                params: {
                    startAt: typeof parameters !== 'string' ? parameters.startAt : undefined,
                    maxResults: typeof parameters !== 'string' ? parameters.maxResults : undefined,
                    orderBy: typeof parameters !== 'string' ? parameters.orderBy : undefined,
                    query: typeof parameters !== 'string' ? parameters.query : undefined,
                    status: typeof parameters !== 'string' ? parameters.status : undefined,
                    expand: typeof parameters !== 'string' ? parameters.expand : undefined,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getProjectVersions(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const projectIdOrKey = typeof parameters === 'string' ? parameters : parameters.projectIdOrKey;
            const config = {
                url: `/rest/api/2/project/${projectIdOrKey}/versions`,
                method: 'GET',
                params: {
                    expand: typeof parameters !== 'string' ? parameters.expand : undefined,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    createVersion(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/version',
                method: 'POST',
                data: {
                    approvers: parameters.approvers,
                    archived: parameters.archived,
                    description: parameters.description,
                    driver: parameters.driver,
                    expand: parameters.expand,
                    id: parameters.id,
                    issuesStatusForFixVersion: parameters.issuesStatusForFixVersion,
                    moveUnfixedIssuesTo: parameters.moveUnfixedIssuesTo,
                    name: parameters.name,
                    operations: parameters.operations,
                    overdue: parameters.overdue,
                    projectId: parameters.projectId,
                    releaseDate: parameters.releaseDate,
                    released: parameters.released,
                    self: parameters.self,
                    startDate: parameters.startDate,
                    userReleaseDate: parameters.userReleaseDate,
                    userStartDate: parameters.userStartDate,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getVersion(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/version/${id}`,
                method: 'GET',
                params: {
                    expand: typeof parameters !== 'string' ? parameters.expand : undefined,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateVersion(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/version/${parameters.id}`,
                method: 'PUT',
                data: {
                    approvers: parameters.approvers,
                    driver: parameters.driver,
                    expand: parameters.expand,
                    description: parameters.description,
                    name: parameters.name,
                    archived: parameters.archived,
                    released: parameters.released,
                    startDate: parameters.startDate,
                    releaseDate: parameters.releaseDate,
                    projectId: parameters.projectId,
                    moveUnfixedIssuesTo: parameters.moveUnfixedIssuesTo,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    mergeVersions(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/version/${parameters.id}/mergeto/${parameters.moveIssuesTo}`,
                method: 'PUT',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    moveVersion(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/version/${parameters.id}/move`,
                method: 'POST',
                data: {
                    after: parameters.after,
                    position: parameters.position,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getVersionRelatedIssues(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/version/${id}/relatedIssueCounts`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getRelatedWork(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/version/${parameters.id}/relatedwork`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    createRelatedWork(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/version/${parameters.id}/relatedwork`,
                method: 'POST',
                data: {
                    category: parameters.category,
                    issueId: parameters.issueId,
                    relatedWorkId: parameters.relatedWorkId,
                    title: parameters.title,
                    url: parameters.url,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateRelatedWork(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/version/${parameters.id}/relatedwork`,
                method: 'PUT',
                data: {
                    category: parameters.category,
                    issueId: parameters.issueId,
                    relatedWorkId: parameters.relatedWorkId,
                    title: parameters.title,
                    url: parameters.url,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteAndReplaceVersion(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/version/${parameters.id}/removeAndSwap`,
                method: 'POST',
                data: {
                    customFieldReplacementList: parameters.customFieldReplacementList,
                    moveAffectedIssuesTo: parameters.moveAffectedIssuesTo,
                    moveFixIssuesTo: parameters.moveFixIssuesTo,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getVersionUnresolvedIssues(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/version/${id}/unresolvedIssueCount`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteRelatedWork(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/version/${parameters.versionId}/relatedwork/${parameters.relatedWorkId}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.ProjectVersions = ProjectVersions;
