// jira.js/dist/version2/projectTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectTypes = void 0;
const tslib_1 = require("tslib");
class ProjectTypes {
    constructor(client) {
        this.client = client;
    }
    getAllProjectTypes(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/project/type',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getAllAccessibleProjectTypes(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/project/type/accessible',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getProjectTypeByKey(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const projectTypeKey = typeof parameters === 'string' ? parameters : parameters.projectTypeKey;
            const config = {
                url: `/rest/api/2/project/type/${projectTypeKey}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getAccessibleProjectTypeByKey(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const projectTypeKey = typeof parameters === 'string' ? parameters : parameters.projectTypeKey;
            const config = {
                url: `/rest/api/2/project/type/${projectTypeKey}/accessible`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.ProjectTypes = ProjectTypes;
