// jira.js/dist/version2/projectClassificationLevels.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectClassificationLevels = void 0;
const tslib_1 = require("tslib");
class ProjectClassificationLevels {
    constructor(client) {
        this.client = client;
    }
    getDefaultProjectClassification(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/project/${parameters.projectIdOrKey}/classification-level/default`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateDefaultProjectClassification(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/project/${parameters.projectIdOrKey}/classification-level/default`,
                method: 'PUT',
                data: {
                    id: parameters.id,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    removeDefaultProjectClassification(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/project/${parameters.projectIdOrKey}/classification-level/default`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.ProjectClassificationLevels = ProjectClassificationLevels;
