// jira.js/dist/version2/projectProperties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectProperties = void 0;
const tslib_1 = require("tslib");
class ProjectProperties {
    constructor(client) {
        this.client = client;
    }
    getProjectPropertyKeys(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const projectIdOrKey = typeof parameters === 'string' ? parameters : parameters.projectIdOrKey;
            const config = {
                url: `/rest/api/2/project/${projectIdOrKey}/properties`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getProjectProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/project/${parameters.projectIdOrKey}/properties/${parameters.propertyKey}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    setProjectProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/project/${parameters.projectIdOrKey}/properties/${parameters.propertyKey}`,
                method: 'PUT',
                data: parameters.propertyValue,
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteProjectProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/project/${parameters.projectIdOrKey}/properties/${parameters.propertyKey}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.ProjectProperties = ProjectProperties;
