// jira.js/dist/version2/projectComponents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectComponents = void 0;
const tslib_1 = require("tslib");
class ProjectComponents {
    constructor(client) {
        this.client = client;
    }
    findComponentsForProjects(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/component',
                method: 'GET',
                params: {
                    projectIdsOrKeys: parameters === null || parameters === void 0 ? void 0 : parameters.projectIdsOrKeys,
                    startAt: parameters === null || parameters === void 0 ? void 0 : parameters.startAt,
                    maxResults: parameters === null || parameters === void 0 ? void 0 : parameters.maxResults,
                    orderBy: parameters === null || parameters === void 0 ? void 0 : parameters.orderBy,
                    query: parameters === null || parameters === void 0 ? void 0 : parameters.query,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    createComponent(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/component',
                method: 'POST',
                data: {
                    ari: parameters.ari,
                    assignee: parameters.assignee,
                    assigneeType: parameters.assigneeType,
                    description: parameters.description,
                    id: parameters.id,
                    isAssigneeTypeValid: parameters.isAssigneeTypeValid,
                    lead: parameters.lead,
                    leadAccountId: parameters.leadAccountId,
                    leadUserName: parameters.leadUserName,
                    metadata: parameters.metadata,
                    name: parameters.name,
                    project: parameters.project,
                    projectId: parameters.projectId,
                    realAssignee: parameters.realAssignee,
                    realAssigneeType: parameters.realAssigneeType,
                    self: parameters.self,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getComponent(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/component/${id}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateComponent(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/component/${parameters.id}`,
                method: 'PUT',
                data: {
                    name: parameters.name,
                    description: parameters.description,
                    leadUserName: parameters.leadUserName,
                    leadAccountId: parameters.leadAccountId,
                    assigneeType: parameters.assigneeType,
                    project: parameters.project,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteComponent(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/component/${id}`,
                method: 'DELETE',
                params: {
                    moveIssuesTo: typeof parameters !== 'string' ? parameters.moveIssuesTo : undefined,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getComponentRelatedIssues(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/component/${id}/relatedIssueCounts`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getProjectComponentsPaginated(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/project/${parameters.projectIdOrKey}/component`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                    orderBy: parameters.orderBy,
                    componentSource: parameters.componentSource,
                    query: parameters.query,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getProjectComponents(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const projectIdOrKey = typeof parameters === 'string' ? parameters : parameters.projectIdOrKey;
            const config = {
                url: `/rest/api/2/project/${projectIdOrKey}/components`,
                method: 'GET',
                params: {
                    componentSource: typeof parameters !== 'string' ? parameters.componentSource : undefined,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.ProjectComponents = ProjectComponents;
