// jira.js/dist/version2/projectFeatures.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectFeatures = void 0;
const tslib_1 = require("tslib");
class ProjectFeatures {
    constructor(client) {
        this.client = client;
    }
    getFeaturesForProject(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const projectIdOrKey = typeof parameters === 'string' ? parameters : parameters.projectIdOrKey;
            const config = {
                url: `/rest/api/2/project/${projectIdOrKey}/features`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    toggleFeatureForProject(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/project/${parameters.projectIdOrKey}/features/${parameters.featureKey}`,
                method: 'PUT',
                data: {
                    state: parameters.state,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.ProjectFeatures = ProjectFeatures;
