// jira.js/dist/version2/screenTabFields.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScreenTabFields = void 0;
const tslib_1 = require("tslib");
class ScreenTabFields {
    constructor(client) {
        this.client = client;
    }
    getAllScreenTabFields(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/screens/${parameters.screenId}/tabs/${parameters.tabId}/fields`,
                method: 'GET',
                params: {
                    projectKey: parameters.projectKey,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    addScreenTabField(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/screens/${parameters.screenId}/tabs/${parameters.tabId}/fields`,
                method: 'POST',
                data: {
                    fieldId: parameters.fieldId,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    removeScreenTabField(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/screens/${parameters.screenId}/tabs/${parameters.tabId}/fields/${parameters.id}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    moveScreenTabField(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/screens/${parameters.screenId}/tabs/${parameters.tabId}/fields/${parameters.id}/move`,
                method: 'POST',
                data: {
                    after: parameters.after,
                    position: parameters.position,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.ScreenTabFields = ScreenTabFields;
