// jira.js/dist/version2/tasks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tasks = void 0;
const tslib_1 = require("tslib");
class Tasks {
    constructor(client) {
        this.client = client;
    }
    getTask(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const taskId = typeof parameters === 'string' ? parameters : parameters.taskId;
            const config = {
                url: `/rest/api/2/task/${taskId}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    cancelTask(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const taskId = typeof parameters === 'string' ? parameters : parameters.taskId;
            const config = {
                url: `/rest/api/2/task/${taskId}/cancel`,
                method: 'POST',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.Tasks = Tasks;
