// jira.js/dist/version2/userProperties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserProperties = void 0;
const tslib_1 = require("tslib");
class UserProperties {
    constructor(client) {
        this.client = client;
    }
    getUserPropertyKeys(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/user/properties',
                method: 'GET',
                params: {
                    accountId: parameters === null || parameters === void 0 ? void 0 : parameters.accountId,
                    userKey: parameters === null || parameters === void 0 ? void 0 : parameters.userKey,
                    username: parameters === null || parameters === void 0 ? void 0 : parameters.username,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getUserProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/user/properties/${parameters.propertyKey}`,
                method: 'GET',
                params: {
                    accountId: parameters.accountId,
                    userKey: parameters.userKey,
                    username: parameters.username,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    setUserProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/user/properties/${parameters.propertyKey}`,
                method: 'PUT',
                params: {
                    accountId: parameters.accountId,
                },
                data: parameters.propertyValue,
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteUserProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/user/properties/${parameters.propertyKey}`,
                method: 'DELETE',
                params: {
                    accountId: parameters.accountId,
                    userKey: parameters.userKey,
                    username: parameters.username,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.UserProperties = UserProperties;
