// jira.js/dist/version2/workflowStatusCategories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowStatusCategories = void 0;
const tslib_1 = require("tslib");
class WorkflowStatusCategories {
    constructor(client) {
        this.client = client;
    }
    getStatusCategories(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/statuscategory',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getStatusCategory(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const idOrKey = typeof parameters === 'string' ? parameters : parameters.idOrKey;
            const config = {
                url: `/rest/api/2/statuscategory/${idOrKey}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.WorkflowStatusCategories = WorkflowStatusCategories;
