// jira.js/dist/version2/workflowSchemeProjectAssociations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowSchemeProjectAssociations = void 0;
const tslib_1 = require("tslib");
class WorkflowSchemeProjectAssociations {
    constructor(client) {
        this.client = client;
    }
    getWorkflowSchemeProjectAssociations(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/workflowscheme/project',
                method: 'GET',
                params: {
                    projectId: parameters.projectId,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    assignSchemeToProject(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/2/workflowscheme/project',
                method: 'PUT',
                data: {
                    workflowSchemeId: parameters.workflowSchemeId,
                    projectId: parameters.projectId,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.WorkflowSchemeProjectAssociations = WorkflowSchemeProjectAssociations;
