// jira.js/dist/version2/workflowSchemeDrafts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowSchemeDrafts = void 0;
const tslib_1 = require("tslib");
class WorkflowSchemeDrafts {
    constructor(client) {
        this.client = client;
    }
    createWorkflowSchemeDraftFromParent(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/workflowscheme/${id}/createdraft`,
                method: 'POST',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getWorkflowSchemeDraft(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/workflowscheme/${id}/draft`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateWorkflowSchemeDraft(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/workflowscheme/${parameters.id}/draft`,
                method: 'PUT',
                data: {
                    name: parameters.name,
                    description: parameters.description,
                    defaultWorkflow: parameters.defaultWorkflow,
                    issueTypeMappings: parameters.issueTypeMappings,
                    updateDraftIfNeeded: parameters.updateDraftIfNeeded,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteWorkflowSchemeDraft(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/workflowscheme/${id}/draft`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getDraftDefaultWorkflow(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/workflowscheme/${id}/draft/default`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateDraftDefaultWorkflow(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/workflowscheme/${parameters.id}/draft/default`,
                method: 'PUT',
                data: {
                    workflow: parameters.workflow,
                    updateDraftIfNeeded: parameters.updateDraftIfNeeded,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteDraftDefaultWorkflow(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/workflowscheme/${id}/draft/default`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getWorkflowSchemeDraftIssueType(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/workflowscheme/${parameters.id}/draft/issuetype/${parameters.issueType}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    setWorkflowSchemeDraftIssueType(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/workflowscheme/${parameters.id}/draft/issuetype/${parameters.issueType}`,
                method: 'PUT',
                data: parameters.details,
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteWorkflowSchemeDraftIssueType(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/workflowscheme/${parameters.id}/draft/issuetype/${parameters.issueType}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    publishDraftWorkflowScheme(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/2/workflowscheme/${id}/draft/publish`,
                method: 'POST',
                params: {
                    validateOnly: typeof parameters !== 'string' && parameters.validateOnly,
                },
                data: {
                    statusMappings: typeof parameters !== 'string' && parameters.statusMappings,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getDraftWorkflow(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/workflowscheme/${parameters.id}/draft/workflow`,
                method: 'GET',
                params: {
                    workflowName: parameters.workflowName,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateDraftWorkflowMapping(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/workflowscheme/${parameters.id}/draft/workflow`,
                method: 'PUT',
                params: {
                    workflowName: parameters.workflowName,
                },
                data: {
                    workflow: parameters.workflow,
                    issueTypes: parameters.issueTypes,
                    defaultMapping: parameters.defaultMapping,
                    updateDraftIfNeeded: parameters.updateDraftIfNeeded,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteDraftWorkflowMapping(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/2/workflowscheme/${parameters.id}/draft/workflow`,
                method: 'DELETE',
                params: {
                    workflowName: parameters.workflowName,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.WorkflowSchemeDrafts = WorkflowSchemeDrafts;
