// jira.js/dist/serviceDesk/client/serviceDeskClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceDeskClient = void 0;
const baseClient_1 = require("../../clients/baseClient");
const customer_1 = require("../customer");
const info_1 = require("../info");
const insight_1 = require("../insight");
const knowledgeBase_1 = require("../knowledgeBase");
const organization_1 = require("../organization");
const request_1 = require("../request");
const requestType_1 = require("../requestType");
const serviceDesk_1 = require("../serviceDesk");
class ServiceDeskClient extends baseClient_1.BaseClient {
    constructor() {
        super(...arguments);
        this.customer = new customer_1.Customer(this);
        this.info = new info_1.Info(this);
        this.insights = new insight_1.Insight(this);
        this.knowledgeBase = new knowledgeBase_1.KnowledgeBase(this);
        this.organization = new organization_1.Organization(this);
        this.request = new request_1.Request(this);
        this.requestType = new requestType_1.RequestType(this);
        this.serviceDesk = new serviceDesk_1.ServiceDesk(this);
    }
}
exports.ServiceDeskClient = ServiceDeskClient;
