// typedoc-plugin-extras/dist/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deprecatedOption = exports.getDateTimeScript = exports.getLastModifiedScript = exports.replaceDescription = exports.replaceTopMostTitle = exports.isUrl = exports.setupNewlineInFooter = exports.appendToFooter = exports.appendFavicon = exports.makeRelativeToRoot = void 0;
/**
 * Creates a relative path from a host file to a target file which is located in the root.
 * @example `modules/_index_.html` --> `../favicon.ico`
 * @param hostPath The path of the file which will contain the resulting relative path.
 * @param targetFilename The name of the target file.
 */
function makeRelativeToRoot(hostPath, targetFilename) {
    // Find separators.
    const match = hostPath.match(/[/\\]/g);
    if (!match)
        return targetFilename;
    // Create path with one '../' per separator and targetFilename at the end.
    const separatorCount = match.length;
    const parts = [...Array(separatorCount).fill('..'), targetFilename];
    return parts.join('/');
}
exports.makeRelativeToRoot = makeRelativeToRoot;
/**
 * Appends a favicon link at the end of `<head>` in the given html.
 * @param html HTML string to append to.
 * @param url URL of the favicon.
 */
function appendFavicon(html, url) {
    return html.replace('</head>', `<link rel="icon" href="${url}" />` +
        '\n' + // Push the end of <head> to the next line.
        '</head>');
}
exports.appendFavicon = appendFavicon;
/**
 * Appends a string value after "Generated using TypeDoc".
 * @param html HTML string to append to.
 * @param value A string value.
 */
function appendToFooter(html, value) {
    return html.replace(/(<p.*>Generated using.*TypeDoc.*)(<\/p>)/, '$1' + // Start of <p>
        value +
        '$2' // End of <p>
    );
}
exports.appendToFooter = appendToFooter;
/**
 * Sets up a space between the first line of the footer and the date.
 * @param html HTML string to append to.
 */
function setupNewlineInFooter(html) {
    return html.replace(/(<p)(>Generated using.*TypeDoc)/, '$1' + // Opening of <p> tag
        ' style="line-height: 28px;"' + // Add spacing between the first line of footer and the date.
        '$2' // End of "Generated using TypeDoc"
    );
}
exports.setupNewlineInFooter = setupNewlineInFooter;
/**
 * Determines whether a string is a URL.
 * @param url The URL to check.
 */
function isUrl(url) {
    return /^https?:\/\//i.test(url);
}
exports.isUrl = isUrl;
/**
 * Replaces the top-most title text.
 * @param html HTML string to replace into.
 * @param title The new title to set.
 */
function replaceTopMostTitle(html, title) {
    return html.replace(/(<a href=")([^"]*)(" class="title">)([^<]*)(<\/a>)/, '$1' + // Start of <a>
        '$2' + // The href link
        '$3' + // The class
        title +
        '$5' // End of <a>
    );
}
exports.replaceTopMostTitle = replaceTopMostTitle;
/**
 * Replaces the meta description
 * @param html HTML string to replace into.
 * @param description The new description to set.
 */
function replaceDescription(html, description) {
    return html.replace(/(<meta name="description" content=")([^<]*)("\/>)/, '$1' + // Start of meta tag
        description.replace('"', '') + // The description (also preventing escaping the attribute)
        '$3' // end of meta tag
    );
}
exports.replaceDescription = replaceDescription;
const getLastModifiedScript = () => {
    // Use English as the locale because we say "Last modified".
    // The title of the element (shown on hover) contains the locale string with the user's timezone.
    return `
        const formatter = new Intl.RelativeTimeFormat('en', {
            numeric: 'auto',
            style: 'short',
        });

        const divisions = [
            { amount: 60, name: 'seconds' },
            { amount: 60, name: 'minutes' },
            { amount: 24, name: 'hours' },
            { amount: 7, name: 'days' },
            { amount: 4.34524, name: 'weeks' },
            { amount: 12, name: 'months' },
            { amount: Number.POSITIVE_INFINITY, name: 'years' }
        ];

        function formatTimeAgo(date) {
            let duration = (date - new Date()) / 1000;
            
            for (const division of divisions) {
                if (Math.abs(duration) < division.amount) {
                    return formatter.format(Math.round(duration), division.name);
                }
                duration /= division.amount;
            }
        }

        document.getElementById('generation-date').title = new Date(window.GENERATION_DATE).toLocaleString();
        document.getElementById('generation-date').innerText = \`Last modified \${formatTimeAgo(window.GENERATION_DATE)}\`;
    `;
};
exports.getLastModifiedScript = getLastModifiedScript;
const getDateTimeScript = (options) => {
    const args = [];
    if (options.footerDate)
        args.push("dateStyle: 'medium'");
    if (options.footerTime)
        args.push("timeStyle: 'long'");
    // Use the browser's language since we just print a date.
    return `
        const formatter = new Intl.DateTimeFormat(navigator.language, {
            ${args.join(',')}
        });

        document.getElementById('generation-date').innerText = formatter.format(window.GENERATION_DATE);
    `;
};
exports.getDateTimeScript = getDateTimeScript;
const deprecatedOption = (app, { name, inFavorOf }) => {
    if (app.options.getValue(name)) {
        const error = Error(`[typedoc-plugin-extras] The \`--${name}\` option is deprecated. Please use \`--${inFavorOf}\` instead.`);
        delete error.stack;
        throw error;
    }
};
exports.deprecatedOption = deprecatedOption;
