// jira.js/dist/version3/announcementBanner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnnouncementBanner = void 0;
const tslib_1 = require("tslib");
class AnnouncementBanner {
    constructor(client) {
        this.client = client;
    }
    getBanner(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/announcementBanner',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    setBanner(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/announcementBanner',
                method: 'PUT',
                data: {
                    isDismissible: parameters === null || parameters === void 0 ? void 0 : parameters.isDismissible,
                    isEnabled: parameters === null || parameters === void 0 ? void 0 : parameters.isEnabled,
                    message: parameters === null || parameters === void 0 ? void 0 : parameters.message,
                    visibility: parameters === null || parameters === void 0 ? void 0 : parameters.visibility,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.AnnouncementBanner = AnnouncementBanner;
