// jira.js/dist/version3/applicationRoles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationRoles = void 0;
const tslib_1 = require("tslib");
class ApplicationRoles {
    constructor(client) {
        this.client = client;
    }
    getAllApplicationRoles(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/applicationrole',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getApplicationRole(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const key = typeof parameters === 'string' ? parameters : parameters.key;
            const config = {
                url: `/rest/api/3/applicationrole/${key}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.ApplicationRoles = ApplicationRoles;
