// jira.js/dist/version3/auditRecords.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuditRecords = void 0;
const tslib_1 = require("tslib");
class AuditRecords {
    constructor(client) {
        this.client = client;
    }
    getAuditRecords(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/auditing/record',
                method: 'GET',
                params: {
                    offset: parameters === null || parameters === void 0 ? void 0 : parameters.offset,
                    limit: parameters === null || parameters === void 0 ? void 0 : parameters.limit,
                    filter: parameters === null || parameters === void 0 ? void 0 : parameters.filter,
                    from: parameters === null || parameters === void 0 ? void 0 : parameters.from,
                    to: parameters === null || parameters === void 0 ? void 0 : parameters.to,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.AuditRecords = AuditRecords;
