// jira.js/dist/version3/groupAndUserPicker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GroupAndUserPicker = void 0;
const tslib_1 = require("tslib");
class GroupAndUserPicker {
    constructor(client) {
        this.client = client;
    }
    findUsersAndGroups(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/groupuserpicker',
                method: 'GET',
                params: {
                    query: parameters.query,
                    maxResults: parameters.maxResults,
                    showAvatar: parameters.showAvatar,
                    fieldId: parameters.fieldId,
                    projectId: parameters.projectId,
                    issueTypeId: parameters.issueTypeId,
                    avatarSize: parameters.avatarSize,
                    caseInsensitive: parameters.caseInsensitive,
                    excludeConnectAddons: parameters.excludeConnectAddons,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.GroupAndUserPicker = GroupAndUserPicker;
