// jira.js/dist/version3/issueCustomFieldConfigurationApps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueCustomFieldConfigurationApps = void 0;
const tslib_1 = require("tslib");
const paramSerializer_1 = require("../paramSerializer");
class IssueCustomFieldConfigurationApps {
    constructor(client) {
        this.client = client;
    }
    getCustomFieldsConfigurations(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/app/field/context/configuration/list',
                method: 'POST',
                params: {
                    id: parameters === null || parameters === void 0 ? void 0 : parameters.id,
                    fieldContextId: (0, paramSerializer_1.paramSerializer)('fieldContextId', parameters === null || parameters === void 0 ? void 0 : parameters.fieldContextId),
                    issueId: parameters === null || parameters === void 0 ? void 0 : parameters.issueId,
                    projectKeyOrId: parameters === null || parameters === void 0 ? void 0 : parameters.projectKeyOrId,
                    issueTypeId: parameters === null || parameters === void 0 ? void 0 : parameters.issueTypeId,
                    startAt: parameters === null || parameters === void 0 ? void 0 : parameters.startAt,
                    maxResults: parameters === null || parameters === void 0 ? void 0 : parameters.maxResults,
                },
                data: {
                    fieldIdsOrKeys: parameters === null || parameters === void 0 ? void 0 : parameters.fieldIdsOrKeys,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getCustomFieldConfiguration(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const fieldIdOrKey = typeof parameters === 'string' ? parameters : parameters.fieldIdOrKey;
            const config = {
                url: `/rest/api/3/app/field/${fieldIdOrKey}/context/configuration`,
                method: 'GET',
                params: {
                    id: typeof parameters !== 'string' && parameters.id,
                    fieldContextId: typeof parameters !== 'string' && parameters.fieldContextId,
                    issueId: typeof parameters !== 'string' && parameters.issueId,
                    projectKeyOrId: typeof parameters !== 'string' && parameters.projectKeyOrId,
                    issueTypeId: typeof parameters !== 'string' && parameters.issueTypeId,
                    startAt: typeof parameters !== 'string' && parameters.startAt,
                    maxResults: typeof parameters !== 'string' && parameters.maxResults,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateCustomFieldConfiguration(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/app/field/${parameters.fieldIdOrKey}/context/configuration`,
                method: 'PUT',
                data: {
                    configurations: parameters.configurations,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.IssueCustomFieldConfigurationApps = IssueCustomFieldConfigurationApps;
