// jira.js/dist/version3/issueFields.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueFields = void 0;
const tslib_1 = require("tslib");
class IssueFields {
    constructor(client) {
        this.client = client;
    }
    getFields(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/field',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    createCustomField(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/field',
                method: 'POST',
                data: {
                    description: parameters.description,
                    name: parameters.name,
                    searcherKey: parameters.searcherKey,
                    type: parameters.type,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getFieldsPaginated(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/field/search',
                method: 'GET',
                params: {
                    startAt: parameters === null || parameters === void 0 ? void 0 : parameters.startAt,
                    maxResults: parameters === null || parameters === void 0 ? void 0 : parameters.maxResults,
                    type: parameters === null || parameters === void 0 ? void 0 : parameters.type,
                    id: parameters === null || parameters === void 0 ? void 0 : parameters.id,
                    query: parameters === null || parameters === void 0 ? void 0 : parameters.query,
                    orderBy: parameters === null || parameters === void 0 ? void 0 : parameters.orderBy,
                    expand: parameters === null || parameters === void 0 ? void 0 : parameters.expand,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getTrashedFieldsPaginated(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/field/search/trashed',
                method: 'GET',
                params: {
                    startAt: parameters === null || parameters === void 0 ? void 0 : parameters.startAt,
                    maxResults: parameters === null || parameters === void 0 ? void 0 : parameters.maxResults,
                    id: parameters === null || parameters === void 0 ? void 0 : parameters.id,
                    query: parameters === null || parameters === void 0 ? void 0 : parameters.query,
                    expand: parameters === null || parameters === void 0 ? void 0 : parameters.expand,
                    orderBy: parameters === null || parameters === void 0 ? void 0 : parameters.orderBy,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateCustomField(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/field/${parameters.fieldId}`,
                method: 'PUT',
                data: {
                    description: parameters.description,
                    name: parameters.name,
                    searcherKey: parameters.searcherKey,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteCustomField(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/field/${parameters.id}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    restoreCustomField(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/field/${parameters.id}/restore`,
                method: 'POST',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    trashCustomField(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/field/${parameters.id}/trash`,
                method: 'POST',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.IssueFields = IssueFields;
