// jira.js/dist/version3/issueCustomFieldValuesApps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueCustomFieldValuesApps = void 0;
const tslib_1 = require("tslib");
class IssueCustomFieldValuesApps {
    constructor(client) {
        this.client = client;
    }
    updateMultipleCustomFieldValues(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/app/field/value',
                method: 'POST',
                params: {
                    generateChangelog: parameters.generateChangelog,
                },
                data: {
                    updates: parameters.updates,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateCustomFieldValue(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/app/field/${parameters.fieldIdOrKey}/value`,
                method: 'PUT',
                params: {
                    generateChangelog: parameters.generateChangelog,
                },
                data: {
                    updates: parameters.updates,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.IssueCustomFieldValuesApps = IssueCustomFieldValuesApps;
