// jira.js/dist/version3/issueSecurityLevel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueSecurityLevel = void 0;
const tslib_1 = require("tslib");
class IssueSecurityLevel {
    constructor(client) {
        this.client = client;
    }
    getIssueSecurityLevelMembers(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/issuesecurityschemes/${parameters.issueSecuritySchemeId}/members`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                    issueSecurityLevelId: parameters.issueSecurityLevelId,
                    expand: parameters.expand,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getIssueSecurityLevel(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/securitylevel/${parameters.id}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.IssueSecurityLevel = IssueSecurityLevel;
