// jira.js/dist/version3/issueVotes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueVotes = void 0;
const tslib_1 = require("tslib");
class IssueVotes {
    constructor(client) {
        this.client = client;
    }
    getVotes(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/issue/${parameters.issueIdOrKey}/votes`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    addVote(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/issue/${parameters.issueIdOrKey}/votes`,
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    removeVote(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/issue/${parameters.issueIdOrKey}/votes`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.IssueVotes = IssueVotes;
