// jira.js/dist/version3/issueLinkTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueLinkTypes = void 0;
const tslib_1 = require("tslib");
class IssueLinkTypes {
    constructor(client) {
        this.client = client;
    }
    getIssueLinkTypes(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/issueLinkType',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    createIssueLinkType(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/issueLinkType',
                method: 'POST',
                data: {
                    id: parameters.id,
                    inward: parameters.inward,
                    name: parameters.name,
                    outward: parameters.outward,
                    self: parameters.self,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getIssueLinkType(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/issueLinkType/${parameters.issueLinkTypeId}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateIssueLinkType(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/issueLinkType/${parameters.issueLinkTypeId}`,
                method: 'PUT',
                data: {
                    id: parameters.id,
                    inward: parameters.inward,
                    name: parameters.name,
                    outward: parameters.outward,
                    self: parameters.self,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteIssueLinkType(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/issueLinkType/${parameters.issueLinkTypeId}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.IssueLinkTypes = IssueLinkTypes;
