// jira.js/dist/version3/jiraExpressions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JiraExpressions = void 0;
const tslib_1 = require("tslib");
class JiraExpressions {
    constructor(client) {
        this.client = client;
    }
    analyseExpression(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/expression/analyse',
                method: 'POST',
                params: {
                    check: parameters === null || parameters === void 0 ? void 0 : parameters.check,
                },
                data: {
                    contextVariables: parameters === null || parameters === void 0 ? void 0 : parameters.contextVariables,
                    expressions: parameters === null || parameters === void 0 ? void 0 : parameters.expressions,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    evaluateJiraExpression(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/expression/eval',
                method: 'POST',
                params: {
                    expand: parameters.expand,
                },
                data: {
                    context: parameters.context,
                    expression: parameters.expression,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    evaluateJiraExpressionUsingEnhancedSearch(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/expression/evaluate',
                method: 'POST',
                params: {
                    expand: parameters.expand,
                },
                data: {
                    expression: parameters.expression,
                    context: parameters.context,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.JiraExpressions = JiraExpressions;
