// jira.js/dist/version3/projectCategories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectCategories = void 0;
const tslib_1 = require("tslib");
class ProjectCategories {
    constructor(client) {
        this.client = client;
    }
    getAllProjectCategories(callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/projectCategory',
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    createProjectCategory(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/projectCategory',
                method: 'POST',
                data: {
                    description: parameters.description,
                    id: parameters.id,
                    name: parameters.name,
                    self: parameters.self,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getProjectCategoryById(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/projectCategory/${parameters.id}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateProjectCategory(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/projectCategory/${parameters.id}`,
                method: 'PUT',
                data: {
                    name: parameters.name,
                    description: parameters.description,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    removeProjectCategory(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/projectCategory/${parameters.id}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.ProjectCategories = ProjectCategories;
