// jira.js/dist/version3/projectEmail.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectEmail = void 0;
const tslib_1 = require("tslib");
class ProjectEmail {
    constructor(client) {
        this.client = client;
    }
    getProjectEmail(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const projectId = typeof parameters === 'string' ? parameters : parameters.projectId;
            const config = {
                url: `/rest/api/3/project/${projectId}/email`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateProjectEmail(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/project/${parameters.projectId}/email`,
                method: 'PUT',
                data: {
                    emailAddress: parameters.emailAddress,
                    emailAddressStatus: parameters.emailAddressStatus,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.ProjectEmail = ProjectEmail;
