// jira.js/dist/version3/screens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Screens = void 0;
const tslib_1 = require("tslib");
class Screens {
    constructor(client) {
        this.client = client;
    }
    getScreensForField(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const fieldId = typeof parameters === 'string' ? parameters : parameters.fieldId;
            const config = {
                url: `/rest/api/3/field/${fieldId}/screens`,
                method: 'GET',
                params: {
                    startAt: typeof parameters !== 'string' && parameters.startAt,
                    maxResults: typeof parameters !== 'string' && parameters.maxResults,
                    expand: typeof parameters !== 'string' && parameters.expand,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getScreens(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/screens',
                method: 'GET',
                params: {
                    startAt: parameters === null || parameters === void 0 ? void 0 : parameters.startAt,
                    maxResults: parameters === null || parameters === void 0 ? void 0 : parameters.maxResults,
                    id: parameters === null || parameters === void 0 ? void 0 : parameters.id,
                    queryString: parameters === null || parameters === void 0 ? void 0 : parameters.queryString,
                    scope: parameters === null || parameters === void 0 ? void 0 : parameters.scope,
                    orderBy: parameters === null || parameters === void 0 ? void 0 : parameters.orderBy,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    createScreen(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/screens',
                method: 'POST',
                data: {
                    description: parameters.description,
                    name: parameters.name,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    addFieldToDefaultScreen(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const fieldId = typeof parameters === 'string' ? parameters : parameters.fieldId;
            const config = {
                url: `/rest/api/3/screens/addToDefault/${fieldId}`,
                method: 'POST',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateScreen(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/screens/${parameters.screenId}`,
                method: 'PUT',
                data: {
                    description: parameters.description,
                    name: parameters.name,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteScreen(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const screenId = typeof parameters === 'string' ? parameters : parameters.screenId;
            const config = {
                url: `/rest/api/3/screens/${screenId}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getAvailableScreenFields(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const screenId = typeof parameters === 'string' ? parameters : parameters.screenId;
            const config = {
                url: `/rest/api/3/screens/${screenId}/availableFields`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.Screens = Screens;
