// jira.js/dist/version3/workflowSchemes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowSchemes = void 0;
const tslib_1 = require("tslib");
class WorkflowSchemes {
    constructor(client) {
        this.client = client;
    }
    getAllWorkflowSchemes(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/workflowscheme',
                method: 'GET',
                params: {
                    startAt: parameters === null || parameters === void 0 ? void 0 : parameters.startAt,
                    maxResults: parameters === null || parameters === void 0 ? void 0 : parameters.maxResults,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    createWorkflowScheme(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/workflowscheme',
                method: 'POST',
                data: {
                    defaultWorkflow: parameters.defaultWorkflow,
                    description: parameters.description,
                    draft: parameters.draft,
                    id: parameters.id,
                    issueTypeMappings: parameters.issueTypeMappings,
                    issueTypes: parameters.issueTypes,
                    lastModified: parameters.lastModified,
                    lastModifiedUser: parameters.lastModifiedUser,
                    name: parameters.name,
                    originalDefaultWorkflow: parameters.originalDefaultWorkflow,
                    originalIssueTypeMappings: parameters.originalIssueTypeMappings,
                    self: parameters.self,
                    updateDraftIfNeeded: parameters.updateDraftIfNeeded,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    readWorkflowSchemes(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/workflowscheme/read',
                method: 'POST',
                data: {
                    projectIds: parameters.projectIds,
                    workflowSchemeIds: parameters.workflowSchemeIds,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateSchemes(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/workflowscheme/update',
                method: 'POST',
                data: {
                    defaultWorkflowId: parameters.defaultWorkflowId,
                    description: parameters.description,
                    id: parameters.id,
                    name: parameters.name,
                    statusMappingsByIssueTypeOverride: parameters.statusMappingsByIssueTypeOverride,
                    statusMappingsByWorkflows: parameters.statusMappingsByWorkflows,
                    version: parameters.version,
                    workflowsForIssueTypes: parameters.workflowsForIssueTypes,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateWorkflowSchemeMappings(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/workflowscheme/update/mappings',
                method: 'POST',
                data: {
                    defaultWorkflowId: parameters.defaultWorkflowId,
                    id: parameters.id,
                    workflowsForIssueTypes: parameters.workflowsForIssueTypes,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getWorkflowScheme(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/3/workflowscheme/${id}`,
                method: 'GET',
                params: {
                    returnDraftIfExists: typeof parameters !== 'string' && parameters.returnDraftIfExists,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateWorkflowScheme(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/workflowscheme/${parameters.id}`,
                method: 'PUT',
                data: {
                    name: parameters.name,
                    description: parameters.description,
                    defaultWorkflow: parameters.defaultWorkflow,
                    issueTypeMappings: parameters.issueTypeMappings,
                    updateDraftIfNeeded: parameters.updateDraftIfNeeded,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteWorkflowScheme(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/3/workflowscheme/${id}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getDefaultWorkflow(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/3/workflowscheme/${id}/default`,
                method: 'GET',
                params: {
                    returnDraftIfExists: typeof parameters !== 'string' && parameters.returnDraftIfExists,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateDefaultWorkflow(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/workflowscheme/${parameters.id}/default`,
                method: 'PUT',
                data: {
                    updateDraftIfNeeded: parameters.updateDraftIfNeeded,
                    workflow: parameters.workflow,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteDefaultWorkflow(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/3/workflowscheme/${id}/default`,
                method: 'DELETE',
                params: {
                    updateDraftIfNeeded: typeof parameters !== 'string' && parameters.updateDraftIfNeeded,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getWorkflowSchemeIssueType(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/workflowscheme/${parameters.id}/issuetype/${parameters.issueType}`,
                method: 'GET',
                params: {
                    returnDraftIfExists: parameters.returnDraftIfExists,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    setWorkflowSchemeIssueType(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/workflowscheme/${parameters.id}/issuetype/${parameters.issueType}`,
                method: 'PUT',
                data: parameters.details,
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteWorkflowSchemeIssueType(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/workflowscheme/${parameters.id}/issuetype/${parameters.issueType}`,
                method: 'DELETE',
                params: {
                    updateDraftIfNeeded: parameters.updateDraftIfNeeded,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getWorkflow(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/3/workflowscheme/${id}/workflow`,
                method: 'GET',
                params: {
                    workflowName: typeof parameters !== 'string' && parameters.workflowName,
                    returnDraftIfExists: typeof parameters !== 'string' && parameters.returnDraftIfExists,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateWorkflowMapping(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/workflowscheme/${parameters.id}/workflow`,
                method: 'PUT',
                params: {
                    workflowName: parameters.workflowName,
                },
                data: {
                    defaultMapping: parameters.defaultMapping,
                    issueTypes: parameters.issueTypes,
                    updateDraftIfNeeded: parameters.updateDraftIfNeeded,
                    workflow: parameters.workflow,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteWorkflowMapping(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const id = typeof parameters === 'string' ? parameters : parameters.id;
            const config = {
                url: `/rest/api/3/workflowscheme/${id}/workflow`,
                method: 'DELETE',
                params: {
                    workflowName: typeof parameters !== 'string' && parameters.workflowName,
                    updateDraftIfNeeded: typeof parameters !== 'string' && parameters.updateDraftIfNeeded,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getProjectUsagesForWorkflowScheme(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/api/3/workflowscheme/${parameters.workflowSchemeId}/projectUsages`,
                method: 'GET',
                params: {
                    nextPageToken: parameters.nextPageToken,
                    maxResults: parameters.maxResults,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.WorkflowSchemes = WorkflowSchemes;
