// jira.js/dist/version3/workflowTransitionRules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowTransitionRules = void 0;
const tslib_1 = require("tslib");
class WorkflowTransitionRules {
    constructor(client) {
        this.client = client;
    }
    getWorkflowTransitionRuleConfigurations(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/workflow/rule/config',
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                    types: parameters.types,
                    keys: parameters.keys,
                    workflowNames: parameters.workflowNames,
                    withTags: parameters.withTags,
                    draft: parameters.draft,
                    expand: parameters.expand,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    updateWorkflowTransitionRuleConfigurations(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/workflow/rule/config',
                method: 'PUT',
                data: {
                    workflows: parameters.workflows,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteWorkflowTransitionRuleConfigurations(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/api/3/workflow/rule/config/delete',
                method: 'PUT',
                data: {
                    workflows: parameters === null || parameters === void 0 ? void 0 : parameters.workflows,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.WorkflowTransitionRules = WorkflowTransitionRules;
