// prettier-plugin-jsdoc/dist/roles.js
import { ABSTRACT, ASYNC, AUGMENTS, AUTHOR, BORROWS, CALLBACK, CATEGORY, CLASS, CONSTANT, DEFAULT, DEFAULT_VALUE, DEPRECATED, DESCRIPTION, EXAMPLE, EXTENDS, EXTERNAL, FILE, FIRES, FLOW, FUNCTION, IGNORE, LICENSE, MEMBER, MEMBEROF, MODULE, NAMESPACE, OVERLOAD, OVERRIDE, PARAM, PRIVATE, PRIVATE_REMARKS, PROPERTY, PROVIDES_MODULE, REMARKS, RETURNS, SEE, SINCE, TEMPLATE, THROWS, TODO, TYPE, SATISFIES, TYPE_PARAM, TYPEDEF, VERSION, YIELDS, } from "./tags.js";
const TAGS_SYNONYMS = {
    arg: PARAM,
    argument: PARAM,
    const: CONSTANT,
    constructor: CLASS,
    desc: DESCRIPTION,
    emits: FIRES,
    examples: EXAMPLE,
    exception: THROWS,
    fileoverview: FILE,
    func: FUNCTION,
    host: EXTERNAL,
    method: FUNCTION,
    overview: FILE,
    params: PARAM,
    prop: PROPERTY,
    return: RETURNS,
    var: MEMBER,
    virtual: ABSTRACT,
    yield: YIELDS,
    hidden: IGNORE,
};
const TAGS_DEFAULT = [DEFAULT, DEFAULT_VALUE];
const TAGS_NAMELESS = [
    BORROWS,
    CATEGORY,
    DEPRECATED,
    DESCRIPTION,
    EXAMPLE,
    EXTENDS,
    LICENSE,
    MODULE,
    NAMESPACE,
    OVERLOAD,
    OVERRIDE,
    PRIVATE_REMARKS,
    REMARKS,
    RETURNS,
    SINCE,
    THROWS,
    TODO,
    YIELDS,
    FILE,
    ...TAGS_DEFAULT,
];
const TAGS_TYPELESS = [
    BORROWS,
    BORROWS,
    DEPRECATED,
    DESCRIPTION,
    EXAMPLE,
    IGNORE,
    LICENSE,
    MODULE,
    NAMESPACE,
    OVERLOAD,
    OVERRIDE,
    PRIVATE_REMARKS,
    REMARKS,
    SINCE,
    TODO,
    FILE,
];
const TAGS_PEV_FORMATE_DESCRIPTION = [
    BORROWS,
    ...TAGS_DEFAULT,
    MEMBEROF,
    MODULE,
    SEE,
];
const TAGS_DESCRIPTION_NEEDED = [
    BORROWS,
    CATEGORY,
    DESCRIPTION,
    EXAMPLE,
    PRIVATE_REMARKS,
    REMARKS,
    SINCE,
    TODO,
];
const TAGS_TYPE_NEEDED = [
    EXTENDS,
    PARAM,
    PROPERTY,
    RETURNS,
    THROWS,
    TYPE,
    SATISFIES,
    TYPEDEF,
    YIELDS,
];
const TAGS_VERTICALLY_ALIGN_ABLE = [
    EXTENDS,
    PARAM,
    PROPERTY,
    RETURNS,
    THROWS,
    TYPE,
    SATISFIES,
    TYPEDEF,
    YIELDS,
];
const TAGS_GROUP_HEAD = [CALLBACK, TYPEDEF];
const TAGS_GROUP_CONDITION = [
    ...TAGS_GROUP_HEAD,
    TYPE,
    PROPERTY,
    PARAM,
    RETURNS,
    YIELDS,
    THROWS,
];
const TAGS_ORDER = {
    [REMARKS]: 1,
    [PRIVATE_REMARKS]: 2,
    [PROVIDES_MODULE]: 3,
    [MODULE]: 4,
    [LICENSE]: 5,
    [FLOW]: 6,
    [ASYNC]: 7,
    [PRIVATE]: 8,
    [IGNORE]: 9,
    [MEMBEROF]: 10,
    [VERSION]: 11,
    [FILE]: 12,
    [AUTHOR]: 13,
    [DEPRECATED]: 14,
    [SINCE]: 15,
    [CATEGORY]: 16,
    [DESCRIPTION]: 17,
    [EXAMPLE]: 18,
    [ABSTRACT]: 19,
    [AUGMENTS]: 20,
    [CONSTANT]: 21,
    [DEFAULT]: 22,
    [DEFAULT_VALUE]: 23,
    [EXTERNAL]: 24,
    [OVERLOAD]: 25,
    [FIRES]: 26,
    [TEMPLATE]: 27,
    [TYPE_PARAM]: 28,
    [FUNCTION]: 29,
    [NAMESPACE]: 30,
    [BORROWS]: 31,
    [CLASS]: 32,
    [EXTENDS]: 33,
    [MEMBER]: 34,
    [TYPEDEF]: 35,
    [TYPE]: 36,
    [SATISFIES]: 37,
    [PROPERTY]: 38,
    [CALLBACK]: 39,
    [PARAM]: 40,
    [YIELDS]: 41,
    [RETURNS]: 42,
    [THROWS]: 43,
    other: 44,
    [SEE]: 45,
    [TODO]: 46,
};
export { TAGS_PEV_FORMATE_DESCRIPTION, TAGS_DESCRIPTION_NEEDED, TAGS_NAMELESS, TAGS_GROUP_HEAD, TAGS_GROUP_CONDITION, TAGS_ORDER, TAGS_SYNONYMS, TAGS_TYPE_NEEDED, TAGS_TYPELESS, TAGS_VERTICALLY_ALIGN_ABLE, TAGS_DEFAULT, };
