// jira.js/dist/agile/board.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Board = void 0;
const tslib_1 = require("tslib");
class Board {
    constructor(client) {
        this.client = client;
    }
    getAllBoards(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/agile/1.0/board',
                method: 'GET',
                params: {
                    startAt: parameters === null || parameters === void 0 ? void 0 : parameters.startAt,
                    maxResults: parameters === null || parameters === void 0 ? void 0 : parameters.maxResults,
                    type: parameters === null || parameters === void 0 ? void 0 : parameters.type,
                    name: parameters === null || parameters === void 0 ? void 0 : parameters.name,
                    projectKeyOrId: parameters === null || parameters === void 0 ? void 0 : parameters.projectKeyOrId,
                    accountIdLocation: parameters === null || parameters === void 0 ? void 0 : parameters.accountIdLocation,
                    projectLocation: parameters === null || parameters === void 0 ? void 0 : parameters.projectLocation,
                    includePrivate: parameters === null || parameters === void 0 ? void 0 : parameters.includePrivate,
                    negateLocationFiltering: parameters === null || parameters === void 0 ? void 0 : parameters.negateLocationFiltering,
                    orderBy: parameters === null || parameters === void 0 ? void 0 : parameters.orderBy,
                    expand: parameters === null || parameters === void 0 ? void 0 : parameters.expand,
                    projectTypeLocation: parameters === null || parameters === void 0 ? void 0 : parameters.projectTypeLocation,
                    filterId: parameters === null || parameters === void 0 ? void 0 : parameters.filterId,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    createBoard(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/agile/1.0/board',
                method: 'POST',
                data: {
                    filterId: parameters.filterId,
                    location: parameters.location,
                    name: parameters.name,
                    type: parameters.type,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getBoardByFilterId(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/filter/${parameters.filterId}`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getBoard(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteBoard(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getIssuesForBacklog(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/backlog`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                    jql: parameters.jql,
                    validateQuery: parameters.validateQuery,
                    fields: parameters.fields,
                    expand: parameters.expand,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getConfiguration(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/configuration`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getEpics(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/epic`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                    done: parameters.done,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getIssuesWithoutEpicForBoard(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/epic/none/issue`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                    jql: parameters.jql,
                    validateQuery: parameters.validateQuery,
                    fields: parameters.fields,
                    expand: parameters.expand,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getBoardIssuesForEpic(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/epic/${parameters.epicId}/issue`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                    jql: parameters.jql,
                    validateQuery: parameters.validateQuery,
                    fields: parameters.fields,
                    expand: parameters.expand,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getFeaturesForBoard(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/features`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    toggleFeatures(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/features`,
                method: 'PUT',
                data: parameters.body,
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getIssuesForBoard(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/issue`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                    jql: parameters.jql,
                    validateQuery: parameters.validateQuery,
                    fields: parameters.fields,
                    expand: parameters.expand,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    moveIssuesToBoard(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/issue`,
                method: 'POST',
                data: {
                    issues: parameters.issues,
                    rankAfterIssue: parameters.rankAfterIssue,
                    rankBeforeIssue: parameters.rankBeforeIssue,
                    rankCustomFieldId: parameters.rankCustomFieldId,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getProjects(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/project`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getProjectsFull(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/project/full`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getBoardPropertyKeys(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/properties`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getBoardProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/properties/${parameters.propertyKey}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    setBoardProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/properties/${parameters.propertyKey}`,
                method: 'PUT',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteBoardProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/properties/${parameters.propertyKey}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getAllQuickFilters(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/quickfilter`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getQuickFilter(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/quickfilter/${parameters.quickFilterId}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getReportsForBoard(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/reports`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getAllSprints(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/sprint`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                    state: parameters.state,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getBoardIssuesForSprint(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/sprint/${parameters.sprintId}/issue`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                    jql: parameters.jql,
                    validateQuery: parameters.validateQuery,
                    fields: parameters.fields,
                    expand: parameters.expand,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getAllVersions(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/board/${parameters.boardId}/version`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                    released: parameters.released,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.Board = Board;
