// jira.js/dist/agile/devopsComponents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DevopsComponents = void 0;
const tslib_1 = require("tslib");
class DevopsComponents {
    constructor(client) {
        this.client = client;
    }
    submitComponents(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/devopscomponents/1.0/bulk',
                method: 'POST',
                data: {
                    properties: parameters.properties,
                    components: parameters.components,
                    providerMetadata: parameters.providerMetadata,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteComponentsByProperty(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/devopscomponents/1.0/bulkByProperties',
                method: 'DELETE',
                params: parameters,
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getComponentById(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/devopscomponents/1.0/${parameters.componentId}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    deleteComponentById(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/devopscomponents/1.0/${parameters.componentId}`,
                method: 'DELETE',
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.DevopsComponents = DevopsComponents;
