// jira.js/dist/agile/epic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Epic = void 0;
const tslib_1 = require("tslib");
class Epic {
    constructor(client) {
        this.client = client;
    }
    getIssuesWithoutEpic(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/agile/1.0/epic/none/issue',
                method: 'GET',
                params: {
                    startAt: parameters === null || parameters === void 0 ? void 0 : parameters.startAt,
                    maxResults: parameters === null || parameters === void 0 ? void 0 : parameters.maxResults,
                    jql: parameters === null || parameters === void 0 ? void 0 : parameters.jql,
                    validateQuery: parameters === null || parameters === void 0 ? void 0 : parameters.validateQuery,
                    fields: parameters === null || parameters === void 0 ? void 0 : parameters.fields,
                    expand: parameters === null || parameters === void 0 ? void 0 : parameters.expand,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    removeIssuesFromEpic(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: '/rest/agile/1.0/epic/none/issue',
                method: 'POST',
                data: {
                    issues: parameters === null || parameters === void 0 ? void 0 : parameters.issues,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getEpic(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/epic/${parameters.epicIdOrKey}`,
                method: 'GET',
            };
            return this.client.sendRequest(config, callback);
        });
    }
    partiallyUpdateEpic(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/epic/${parameters.epicIdOrKey}`,
                method: 'POST',
                data: {
                    color: parameters.color,
                    done: parameters.done,
                    name: parameters.name,
                    summary: parameters.summary,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    getIssuesForEpic(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/epic/${parameters.epicIdOrKey}/issue`,
                method: 'GET',
                params: {
                    startAt: parameters.startAt,
                    maxResults: parameters.maxResults,
                    jql: parameters.jql,
                    validateQuery: parameters.validateQuery,
                    fields: parameters.fields,
                    expand: parameters.expand,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    moveIssuesToEpic(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/epic/${parameters.epicIdOrKey}/issue`,
                method: 'POST',
                data: {
                    issues: parameters.issues,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
    rankEpics(parameters, callback) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const config = {
                url: `/rest/agile/1.0/epic/${parameters.epicIdOrKey}/rank`,
                method: 'PUT',
                data: {
                    rankAfterEpic: parameters.rankAfterEpic,
                    rankBeforeEpic: parameters.rankBeforeEpic,
                    rankCustomFieldId: parameters.rankCustomFieldId,
                },
            };
            return this.client.sendRequest(config, callback);
        });
    }
}
exports.Epic = Epic;
